crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4184,6 +4193,139 @@ impl<'db> TypeInferenceBuilder<'db> {
         Ok(builder.build())
     }
 
+    fn try_infer_type_api_operation(
+        &mut self,
+        api_type: Type<'db>,
+        arguments: &ast::Expr,
+    ) -> Option<Type<'db>> {
+        match api_type.into_class_literal() {
+            Some(class)
+                if file_to_module(self.db(), class.class.file(self.db()))
+                    .is_some_and(|module| module.is_known(KnownModule::RedKnot)) =>
+            {
+                let db = self.db();
+
+                let is_type_of = class.class.name(db).as_str() == "TypeOf";
+
+                let argument_types = match arguments {
+                    ast::Expr::Tuple(tuple) => Either::Left(
+                        tuple
+                            .iter()
+                            .map(|element| self.infer_type_expression(element)),
+                    ),
+                    expr => Either::Right(std::iter::once(if is_type_of {
+                        self.infer_expression(expr)
+                    } else {
+                        self.infer_type_expression(expr)
+                    })),
+                };
+
+                let result = type_api::resolve_type_operation(db, class.class, argument_types);
+
+                match result {
+                    Ok(ty) => ty,
+                    Err(TypeApiArgumentsError { expected, actual }) => {
+                        self.context.report_lint(
+                            &TYPE_API_WRONG_ARITY,
+                            arguments.into(),
+                            format_args!(
+                                "Expected {expected} type argument{}, got {actual}",
+                                if expected == 1 { "" } else { "s" },
+                            ),
+                        );
+
+                        Some(Type::Unknown)
+                    }
+                }
+            }
+            _ => None,
+        }
+    }
+
+    fn try_infer_type_api_predicate(
+        &mut self,
+        function: Type<'db>,
+        arguments: &ast::Arguments,
+    ) -> Option<Type<'db>> {
+        let db = self.db();
+
+        match function.into_function_literal() {
+            Some(function)
+                if file_to_module(self.db(), function.body_scope(self.db()).file(self.db()))
```

Comment:
nit: you assigned `self.db()` to a variable just a couple of lines above ;)

```suggestion
                if file_to_module(db, function.body_scope(db).file(db))
```